hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/transport/server/ratis/ContainerStateMachine.java

```diff
@@ -962,6 +982,11 @@ private CompletableFuture<ContainerCommandResponseProto> applyTransaction(
     final CheckedSupplier<ContainerCommandResponseProto, Exception> task
         = () -> {
           try {
+            try {
+              this.validatePeers(this.ratisServer.getServer(), getGroupId());
+            } catch (StorageContainerException e) {
+              return ContainerUtils.logAndReturnError(LOG, e, request);
+            }
```

Comment:
What is the expectation?  If the peer is not in the group, fail it?

If it is the case, we should do it earlier.  Why not doing it in `startTransaction(LogEntryProto, RaftPeerRole)`?

Reply:
The datanode proposes to close the pipeline to SCM only on the applyTransaction block. Correct me if I am wrong here.

Reply:
https://github.com/apache/ozone/blob/aa760ce9fd3b5823e535bed1e88d5c391bab7ad7/hadoop-hdds/container-service/src/main/java/org/apache/hadoop/ozone/container/common/transport/server/ratis/ContainerStateMachine.java#L1104

Reply:
Just call it in the `validatePeers` method then.

Reply:
We would want to close the entire raft group. If we do it in the start transaction, just the add log will fail on one server right? We would be just performing this check on the leader right? Correct me if I am wrong here. We need this block running on followers as well.

Reply:
> ...  just the add log will fail on one server right? 

`handleApplyTransactionFailure` will call `triggerPipelineClose`.  It will close the pipeline.

> ... We would be just performing this check on the leader right? 

No, `startTransaction(LogEntryProto, RaftPeerRole)` is for follower.  Leader has a different `startTransaction` method.  Please see the javadoc.

We don't have to check leader -- it is not in the group, it won't be the leader.

Reply:
We have to perform the check on both the leader and follower

Reply:
If you really want to  check leader, check it also in the other `startTransaction` method
```java
  TransactionContext startTransaction(RaftClientRequest request) throws IOException;
```

Reply:
Do we set the exception in the transactionContext?

Reply:
Discussed offline: 
We came to a conclusion adding a validation block at applyTransactionSerial & startTransaction would not help. Start transaction on follower doesn't allow throwing an exception and it will bring down the state machine. If the state machine gets closed, we will never be able to clear up the raft meta data after the state machine goes down. The same is true applyTransactionSerial as well. Even though it allows throwing an exception that also brings down the statemachine. So the only easy solution (Handling things on Ozone side and not relying on raft) here is to add it to the apply transaction block and return an error response which would make the containerStateMachine unhealthy but would still allow a close pipeline on it since the statemachine is still healthy.